### PR TITLE
Random Feature Entr Detr Local Closure

### PR DESCRIPTION
**Pull request by nickhnelsen**
## In this PR

- A simple Random Fourier Feature approximation to a scalar-valued GP with squared exponential kernel and four inverse lengthscales (one for each Pi group input)
- Separates optimizable (expansion coefficients and hyperparameters) from fixed (realizations of random variables) parameters in the namelist

### CODE REVIEW

File: src/closures/nondimensional_exchange_functions.jl [hunk 1]
```diff
@@ -193,6 +193,41 @@ function non_dimensional_function(param_set, εδ_model_vars, ::LinearEntr)
...
+    scale_x_entr = abs.(c_rf_opt[1,m+2:m+d+1]).*nondim_groups
+    scale_x_detr = abs.(c_rf_opt[2,m+2:m+d+1]).*nondim_groups
+
+    # Features
+    f_entr = c_rf_opt[1,m+1]*sqrt(2)*cos.(c_rf_fix[1,:,2:d+1]*scale_x_entr + c_rf_fix[1,:,1])
+    f_detr = c_rf_opt[2,m+1]*sqrt(2)*cos.(c_rf_fix[2,:,2:d+1]*scale_x_detr + c_rf_fix[2,:,1])
+
+    # Predict and square output for nonnegativity
+    nondim_ε = sum(c_rf_opt[1,1:m].*f_entr)/m
+    nondim_δ = sum(c_rf_opt[2,1:m].*f_detr)/m
```
Comment (charleskawczynski) [inline src/closures/nondimensional_exchange_functions.jl:227]: We could write this in a non-allocating way, but I think this is fine for now as we're experimenting with NN parameterizations.

**Review by charleskawczynski — APPROVED** [review]
LGTM.

I'm not 100% sure, but I think you may need to extend [this](https://github.com/CliMA/TurbulenceConvection.jl/blob/main/driver/main.jl#L257-L260) (if we're going to run calibration on this.). cc @haakon-e for input

**Comment by nickhnelsen** [comment]
> LGTM.
> 
> I'm not 100% sure, but I think you may need to extend [this](https://github.com/CliMA/TurbulenceConvection.jl/blob/main/driver/main.jl#L257-L260) (if we're going to run calibration on this.). cc @haakon-e for input
> 
> Also, let's squash / rebase before merging.

Thanks for catching that, I'll defer to @haakon-e (in particular, would I only include the parameters here that need to be calibrated?).

Also, I have never squashed/rebased before so any advice on how this is usually done would be appreciated!

**Comment by haakon-e** [comment]
> LGTM.
> 
> I'm not 100% sure, but I think you may need to extend [this](https://github.com/CliMA/TurbulenceConvection.jl/blob/main/driver/main.jl#L257-L260) (if we're going to run calibration on this.). cc @haakon-e for input
> 
> Also, let's squash / rebase before merging.

yes, that needs to be extended.

I'm not super happy about creating new sets of parameter names for every closure we experiment with (that's why we created `c_gen`), but others (@yairchn, @costachris) may have had more thorough discussions on this lately to conclude that we can keep expanding the namelist file.

File: driver/generate_namelist.jl [hunk 2]
```diff
@@ -110,6 +110,24 @@ function default_namelist(case_name::String; root::String = ".", write::Bool = t
...
+                    ones(2,1)/1e2,
+                    ones(2,2), dims=2))
+        )
+
+    # RF: fixed realizations of random variables, 2 x m x (1 + d)
+    namelist_defaults["turbulence"]["EDMF_PrognosticTKE"]["rf_fix_ent_params"] =
+        SA.SVector{1000}(
+        vec(cat(2*pi*rand(2,100,1),
+                    randn(2,100,4), dims=3))
+        )
```
[the remark this answers is not in the document]
Comment (charleskawczynski) [reply driver/generate_namelist.jl:129]: +1 for fixing the seed!

[the remark this answers is not in the document]
Comment (charleskawczynski) [reply driver/generate_namelist.jl:129]: I believe that must be done at the top level, though, so we'll probably want to add that to `driver/main.jl`

[the remark this answers is not in the document]
Comment (haakon-e) [reply driver/generate_namelist.jl:129]: Could we add a random seed as an optional entry to the `namelist`? This way, you can choose whether to fix or not fix the seed e.g. for every EKI iteration.

**Comment by costachris** [comment]
> > LGTM.
> > I'm not 100% sure, but I think you may need to extend [this](https://github.com/CliMA/TurbulenceConvection.jl/blob/main/driver/main.jl#L257-L260) (if we're going to run calibration on this.). cc @haakon-e for input
> > Also, let's squash / rebase before merging.
> 
> yes, that needs to be extended.
> 
> I'm not super happy about creating new sets of parameter names for every closure we experiment with (that's why we created `c_gen`), but others (@yairchn, @costachris) may have had more thorough discussions on this lately to conclude that we can keep expanding the namelist file.

We discussed it but never fully settled on a solution. I agree it's not ideal to have a different namelist vector for every closure type. It seems okay for now since there's only a few closure types and this makes it easy to switch between them, although one solution is to nest vectors under `c_gen` so we only have to pass around `c_gen` through the source code (ie.  `namelist_defaults["turbulence"]["EDMF_PrognosticTKE"]["c_gen"]["rf_fix_ent_params"]`... and similarly for others

File: src/closures/nondimensional_exchange_functions.jl [hunk 3]
```diff
@@ -193,6 +193,41 @@ function non_dimensional_function(param_set, εδ_model_vars, ::LinearEntr)
...
+    non_dimensional_function(param_set, εδ_model_vars, ::RFEntr)
+
+Uses a Random Feature model to predict the non-dimensional components of dynamical entrainment/detrainment.
+ - `param_set`      :: parameter set
+ - `εδ_model_vars`  :: structure containing variables
+ - `εδ_model_type`  :: RFEntr - basic RF entrainment closure
+"""
+function non_dimensional_function(param_set, εδ_model_vars, ::RFEntr)
+    # d=4 inputs, p=2 outputs, m=100 random features
+    d = 4
```
Comment (yairchn) [inline src/closures/nondimensional_exchange_functions.jl:207]: do we wanna make  `d` and `m` flexible (i.e. move them to namelist arguments)?

File: src/closures/nondimensional_exchange_functions.jl (hunk 1 above)
Comment (yairchn) [reply src/closures/nondimensional_exchange_functions.jl:227]: if the non-allocating way simple - maybe it canoe done now and not worry about it in the future ?

**Review by yairchn — APPROVED** [review]
I left a few comments but overall looks good to me

File: src/closures/nondimensional_exchange_functions.jl (hunk 1 above)
Comment (charleskawczynski) [reply src/closures/nondimensional_exchange_functions.jl:227]: I think it's probably fine for now while we're experimenting. It's important that people using these untested (via the test suite) experimental features report that they're slow (if they are) though.

File: driver/generate_namelist.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (nickhnelsen) [reply driver/generate_namelist.jl:129]: Thanks, I'll add this

File: src/closures/nondimensional_exchange_functions.jl (hunk 3 above)
Comment (nickhnelsen) [reply src/closures/nondimensional_exchange_functions.jl:207]: Yes, m should be flexible. I'll add this

**Comment by charleskawczynski** [comment]
bors try

**Comment by yairchn** [comment]
> There are some performance issues we should figure out how to manage before merging. I think it may have to due with the matrices in the parameter set.

@charleskawczynski is there something we can do to help them resolve these issue and move forward with the PR?

**Comment by charleskawczynski** [comment]
Can we try using regular vectors instead of SVectors in the namelist and lift the `isbits` limitation? (just add a separate commit and see if CI is faster)

File: driver/main.jl [hunk 4]
```diff
@@ -254,7 +254,13 @@ nc_results_file(::Nothing) = @info "The simulation was run without IO, so no nc
 
 function main1d(namelist; time_run = true)
     # TODO: generalize conversion of arrays from namelist to `SVector`s.
-    for param_name in ["general_ent_params", "general_stochastic_ent_params", "fno_ent_params"]
+    for param_name in [
+        "general_ent_params",
+        "general_stochastic_ent_params",
+        "fno_ent_params",
+        "rf_opt_ent_params",
+        "rf_fix_ent_params",
```
Comment (nickhnelsen) [inline driver/main.jl:262]: comma should not be on line 262, will remove after performance test without SVec

**Comment by yairchn** [comment]
@charleskawczynski can you take one more look at the buildkite before we merge?

**Review by charleskawczynski — APPROVED** [review]

**Pull request merged**